### PR TITLE
Allow custom filters for LDAP group search

### LINKED ISSUE #11697: about GroupOfUniqueNames and groupOfNames
When I use GroupOfUniqueNames, I can't achieve the mapping effect。But using GroupOfNames is OK. Can you also support GroupOfUniqueNames!
image  
This is the log of LDAP！

### LINKED ISSUE #12151: LDAP Group Sync Broken (Docker, v0.34.3)
**Describe the bug**
LDAP group sync is not working. Groups are not synced at all. I have tried to enable group sync, logging in and out; even try with new account. Not only groups don't get synced, users are also get removed from their current non-LDAP group.

**Logs**
I don't see any abnormal logs from the service container

**To Reproduce**
1. Go to 'Admin' > 'Setting' > 'Authenication' > LDAP 'Configure'
2. Enable `SYNCHRONIZE GROUP MEMBERSHIPS`
3. Set `GROUP SEARCH BASE` to `cn=groups,dc=mycompany,dc=com`
4. User A logs out and logs back in
5. No LDAP group created, user A gets removed from all non-LDAP groups

**Expected behavior**
- All LDAP groups that user A belongs to should be created
- User A should be added to those LDAP groups
- User A should not be removed from his current non-LDAP group

**Information about your Metabase Installation:**
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.3 Safari/605.1.15",
    "vendor": "Apple Computer, Inc."
  },
  "system-info": {
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.5+10",
    "java.vendor": "AdoptOpenJDK",
    "java.vendor.url": "https://adoptopenjdk.net/",
    "java.version": "11.0.5",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.5+10",
    "os.name": "Linux",
    "os.version": "4.15.0-42-generic",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "postgres"
    ],
    "hosting-env": "unknown",
    "application-database": "h2",
    "application-database-details": {
      "database": {
        "name": "H2",
        "version": "1.4.197 (2018-03-18)"
      },
      "jdbc-driver": {
        "name": "H2 JDBC Driver",
        "version": "1.4.197 (2018-03-18)"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2020-02-25",
      "tag": "v0.34.3",
      "branch": "release-0.34.x",
      "hash": "1a83edb"
    },
    "settings": {
      "report-timezone": "US/Pacific"
    }
  }
}

**Severity**
We currently have to manually manage user and group permissions, which is a real real pain.

**Additional context**
Here are some examples of our `ldapsearch` results:
1. Search for `user_a`
```bash
$ ldapsearch -x -D "uid=admin,cn=users,dc=mycompany,dc=com" -W -H ldap://ldap.mycompany.com -b "cn=users,dc=mycompany,dc=com" -s sub 'uid=user_a'
# extended LDIF
#
# LDAPv3
# base <cn=users,dc=mycompany,dc=com> with scope subtree
# filter: uid=user_a
# requesting: ALL
#

# anhminh_vu, users, mycompany.com
dn: uid=user_a,cn=users,dc=mycompany,dc=com
uid: user_a
mail: user_a@mycompany.com
sn: foo
uidNumber: 0123456
objectClass: person
givenName: bar
cn: User A
gidNumber: 12345

# search result
search: 2
result: 0 Success

# numResponses: 2
# numEntries: 1
```
2. Search for `mygroup` to which `user_a` belongs
```bash
$ ldapsearch -x -D "uid=admin,cn=users,dc=mycompany,dc=com" -W -H ldap://ldap.mycompany.com -b "cn=groups,dc=mycompany,dc=com" -s sub "cn=mygroup"

# extended LDIF
#
# LDAPv3
# base <cn=groups,dc=mycompany,dc=com> with scope subtree
# filter: cn=mygroup
# requesting: ALL
#

# mygroup, groups, mycompany.com
dn: cn=mygroup,cn=groups,dc=mycompany,dc=com
cn: mygroup
gidNumber: 876543
objectClass: posixGroup
memberUid: user_a
memberUid: user_b

# search result
search: 2
result: 0 Success

# numResponses: 2
# numEntries: 1
```

### LINKED ISSUE #13215: LDAP Group Check with non-default membership keys
**Is your feature request related to a problem? Please describe.**
To use LDAP group sync feature, your LDAP system needs to store members of a group with the key 'member', but as my case LDAP system gives members with the key 'uniqueMember'. Hence I cannot use group sync feature.

**Describe the solution you'd like**
In the code below, member keyword can be parametric and defined by settings UI page.
https://github.com/metabase/metabase/blob/f03f3dfec082460510fef22c00c11cb3e035260a/src/metabase/integrations/ldap.clj#L131-L138

And also in the code below, memberof keyword can be parametric and defined by settings UI page.
https://github.com/metabase/metabase/blob/f03f3dfec082460510fef22c00c11cb3e035260a/src/metabase/integrations/ldap.clj#L186-L205

**How important is this feature to you?**
It is very important for big companies to use ldap groups for authorization.

### PR DESCRIPTION
**Pull request by noahmoss**
This PR builds on the discussion and code from #7098 to support both rfc2307 and rfc2307bis schemas for group membership. In that PR, two new settings were exposed: `ldap-group-schema` (e.g. `member` or `memberUid`) and `ldap-attribute-uid` (e.g. `uid` or `dn`).

I thought would be simpler to just make the filter used to search for group membership customizable in a similar way as the user filter, with placeholders `{dn}` or `{uid}` which are replaced by the respective values returned by the LDAP server for the user. This is also might provide a bit more flexibility for some edge cases of non-standard directory schemas.

I've also adjusted some of the other LDAP settings descriptions for clarity/uniformity.

Here's what the LDAP settings page looks like now: 

image

Fixes #13215 and #12151 and (I think) #11697

### CODE REVIEW

File: src/metabase/integrations/ldap/default_implementation.clj [hunk 1]
```diff
@@ -33,23 +32,34 @@
              :size-limit 1}))
           u/lower-case-map-keys))
 
+(s/defn ^:private process-group-membership-filter :- su/NonBlankString
+  "Replace DN and UID placeholders with values returned by the LDAP server."
+  [group-membership-filter :- su/NonBlankString
+   dn                      :- su/NonBlankString
+   uid                     :- (s/maybe su/NonBlankString)]
+  (let [uid-string (or uid "")]
```
Comment (noahmoss) [inline src/metabase/integrations/ldap/default_implementation.clj:40]: This seems awkward, is there any alternative? I don't think we can assume `uid` is set for a user, so we should default to an empty string if it's not, but doing this inline on L43 results in a reflection warning.

Comment (camsaul) [reply src/metabase/integrations/ldap/default_implementation.clj:40]: If you put `^String` in front of the `(or ...)` it _should_ fix the reflection warning e.g.

```clj
(Filter/encodeValue ^String (or uid ""))
```

File: src/metabase/integrations/ldap/default_implementation.clj [hunk 2]
```diff
@@ -33,23 +32,34 @@
...
                    {:scope  :sub
-                    :filter (Filter/createEqualityFilter "member" ^String dn)})]
+                    :filter (process-group-membership-filter group-membership-filter dn uid)})]
       (map :dn results))))
 
 (s/defn ldap-search-result->user-info :- (s/maybe i/UserInfo)
   "Convert the result "
   [ldap-connection          :- LDAPConnectionPool
-   {:keys [dn], :as result} :- su/Map
+   {:keys [dn, uid], :as result} :- su/Map
```
Comment (camsaul) [inline src/metabase/integrations/ldap/default_implementation.clj:62]: This is so trivial you shouldn't bother fixing it, but just a gentle reminder https://guide.clojure.style/#no-commas-for-seq-literals

File: test/metabase/test/integrations/ldap.clj [hunk 3]
```diff
@@ -13,7 +13,9 @@
 (defn- get-server-config []
   (doto (InMemoryDirectoryServerConfig. (into-array String ["dc=metabase,dc=com"]))
     (.addAdditionalBindCredentials "cn=Directory Manager" "password")
-    (.setSchema (Schema/getDefaultStandardSchema))
+     (.setSchema (Schema/mergeSchemas
+                  (into-array Schema [(Schema/getDefaultStandardSchema)
+                                      (Schema/getSchema [(io/file "test_resources/posixGroup.schema.ldif")])])))
```
Comment (camsaul) [inline test/metabase/test/integrations/ldap.clj:18]: Looks like this is indented one space too many

**Review by camsaul — COMMENTED** [review]
Overall this looks good to me, but I have a couple questions we should get answers for before merging

1.  Do we need to add documentation about the new LDAP Setting? It doesn't need to be done as part of this PR, but we should make sure we don't forget about doing it if we need it. Might want to check with @jeff-bruemmer RE dox
2. Should this be an EE-only feature? (Would that be feasible?) LDAP _would_ have been an EE-only feature like our other SSO integrations, but it was contributed by the community before we launched EE. So our general policy has been maintaining the community-contributed LDAP integration in OSS and offering a more-fully-featured LDAP integration in EE. We do bugfixes for both OSS & EE, but for new features we usually only add them to EE LDAP. @salsakran do you have any opinions on this feature in particular?

File: src/metabase/integrations/ldap/default_implementation.clj (hunk 1 above)
Comment (noahmoss) [reply src/metabase/integrations/ldap/default_implementation.clj:40]: That's what I had when I was seeing the reflection warning :/

**Comment by noahmoss** [comment]
Changed the setting to be EE-only, with the group filter passed in as an argument to `ldap-search-result->user-info` from the active implementation.

Re: docs, that's a good point, currently the [SSO docs](https://www.metabase.com/docs/latest/administration-guide/10-single-sign-on.html) don't go into detail about most of the settings, particularly the EE-only ones, but also things like `User Filter`. Not sure where exactly the more advanced details should go. I can open a new issue to expand the LDAP docs, cc @jeff-bruemmer

File: src/metabase/integrations/ldap/default_implementation.clj (hunk 1 above)
Comment (noahmoss) [reply src/metabase/integrations/ldap/default_implementation.clj:43]: Can confirm, tried this again and the reflection linter prints:
```
Reflection warning, metabase/integrations/ldap/default_implementation.clj:46:30 - call to static method encodeValue on com.unboundid.ldap.sdk.Filter can't be resolved (argument types: unknown).
```

**Review by camsaul — APPROVED** [review]

**Pull request merged**